README.md
# EctoAutoslugField

[![Build Status](https://github.com/sobolevn/ecto_autoslug_field/workflows/test/badge.svg?branch=master&event=push)](https://github.com/sobolevn/ecto_autoslug_field/actions?query=workflow%3Atest)
[![Coverage Status](https://coveralls.io/repos/github/sobolevn/ecto_autoslug_field/badge.svg?branch=master)](https://coveralls.io/github/sobolevn/ecto_autoslug_field?branch=master)
[![Module Version](https://img.shields.io/hexpm/v/ecto_autoslug_field.svg)](https://hex.pm/packages/ecto_autoslug_field)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/ecto_autoslug_field/)
[![License](https://img.shields.io/hexpm/l/ecto_autoslug_field.svg)](https://github.com/sobolevn/ecto_autoslug_field/blob/master/LICENSE)

`ecto_autoslug_field` is a reusable [`Ecto`](https://github.com/elixir-ecto/ecto) library which can automatically create slugs from other fields. We use [`slugify`](https://github.com/jayjun/slugify) as a default slug-engine.

We only depend on the `ecto` package (we do not deal with `ecto_sql` at all).
We support `ecto >= 3.7 and ecto < 4`!

See [this blog post](https://sobolevn.me/2017/07/creating-slugs-for-ecto-schemas)
for more information.


## Installation

```elixir
def deps do
  [
    {:ecto_autoslug_field, "~> 3.1"}
  ]
end
```


## Options

There are several options to configure.

Required:

- `:to` - represents the slug field name where to save value to

Optional:

- `:from` - represents the source fields from which to build slug, if this option is not set you have to override `get_sources/2` function
- `:always_change` - if this option is set slug will be recreated from the given sources each time `maybe_generate_slug` function is called


## Functions

- `get_sources/2` - this function is used to get sources for the slug, [docs](https://hexdocs.pm/ecto_autoslug_field/EctoAutoslugField.SlugBase.html#get_sources/2).
- `build_slug/2` - this function is a place to modify the result slug, [docs](https://hexdocs.pm/ecto_autoslug_field/EctoAutoslugField.SlugBase.html#build_slug/2).


## Examples

The simplest example:

```elixir
defmodule EctoSlugs.Blog.Article.TitleSlug do
  use EctoAutoslugField.Slug, from: :title, to: :slug
end

defmodule EctoSlugs.Blog.Article do
  use Ecto.Schema
  import Ecto.Changeset
  alias EctoSlugs.Blog.Article
  alias EctoSlugs.Blog.Article.TitleSlug

  schema "blog_articles" do
    field :breaking, :boolean, default: false
    field :content, :string
    field :title, :string

    field :slug, TitleSlug.Type

    timestamps()
  end

  def changeset(model, params \\ :invalid) do
    model
    |> cast(params, [:title, :content, :breaking])
    |> validate_required([:title, :content])
    |> unique_constraint(:title)
    |> TitleSlug.maybe_generate_slug()
    |> TitleSlug.unique_constraint()
  end
end
```

See [this tutorial](https://sobolevn.me/2017/07/creating-slugs-for-ecto-schemas)
for some more examples.


## Changelog

See [CHANGELOG.md](./CHANGELOG.md).


## Copyright and License

Copyright (c) 2016 Nikita Sobolev

This library is released under the MIT License. See the [LICENSE.md](./LICENSE.md) file
for further details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## v3.1.0 - 2023-08-10

### Features

- Now supporting `elixir>=1.13`
- Now using `slugify` instead of `slugger`


## v3.0.0 - 2021-09-12

### Features

- Now support `elixir>=1.10`
- Updates `ecto` to `3.7`
- Updates bunch of other deps to newer versions

### Misc

- Moves from Travis to GitHub Actions
- Adds `mix format` support


## v2.0.1 - 2019-10-16

### Bugfixes

- Fixed warnings with `ecto>=3.2`, see [#101](https://github.com/sobolevn/ecto_autoslug_field/pull/101)


## v2.0.0 - 2019-07-08

Breaking features:

- Now supports `elixir>=1.6`
- Now support only `ecto>=3.1`

### Improvements

- Adds `mix format`


## v1.0.0 - 2018-12-28

### Improvements

- Adds support for `ecto >= 3.0`
- Adds official support for `elixir == 1.6` and `elixir == 1.7`


## v0.5.1 - 2018-05-21

### Documentation

- Updates docs on using `force_update_slug`


## v0.5.0 - 2018-05-21

### Improvements

- Adds `force_generate_slug` function to the client API. This function may be used when the `always_change` option of the slug is maybe not set to `true` but you want to override this setting on the given changeset and force regeneration of the slug
- Updates multiple dependencies


## v0.4.0 - 2017-11-23

### Improvements

- Adds support for [numeric, datetime and date fields](https://github.com/sobolevn/ecto_autoslug_field/pull/18)
- Updates dependencies, now supports `ecto >= 2.1`

### Documentation

- Updates `README.md` with the new `ecto` version

### Testing

- Adds new test cases to cover new field types


## v0.3.1 - 2017-07-19

### Improvements

- Adds `build_slug/2` to accept the original `changeset` as the second argument, it still receives list of `sources` as the first argument
- Updates `build_slug/1` inner logic

### Documentation

- Updates `README.md` with the new example
- Updates docs to handle new changes
- Updates `CONTRIBUTING.md` with 'Development' section

### Testing

- Adds new test cases to cover `build_slug/2`


## v0.3.0 - 2017-06-11

### Changes

- *breaking* Updates `mix.exs` with new dependencies, closes [#14](https://github.com/sobolevn/ecto_autoslug_field/issues/14)
- *breaking* Drops `elixir` version 1.2 and `otp` version 17 support

### Improvements

- Updates `cast/4` functions to `cast/3` functions (new Ecto)
- Updates `credo` version and `.credo.exs` config
- Updates `.travis.yml` with credo and new `otp` release

### Bugs

- Fixes dialyxer issues, closes [#10](https://github.com/sobolevn/ecto_autoslug_field/issues/10)

### Documentation

- Updates docs, fixes spelling


## v0.2.1 - 2017-02-01

- Added `credo` support
- Added `elixir` version 1.4 support


## v0.2.0 - 2016-11-02

- Since this version only `ecto` 2 and above are supported
- Updated docs on how to use this package with older `ecto`
- Credo is not working with this release, please see https://github.com/sobolevn/ecto_autoslug_field/issues/5


## v0.1.3 - 2016-07-22

- Now `build_slug/1` is called only when needed, changed docs appropriately
- Replaced `cond` inside `do_generate_slug/3` with more obvious `if`
- Changed tests structure, reached 100% coverage


## v0.1.2 - 2016-07-21

- Moved `get_sources/2` and `build_slug/1` definitions into `SlugBase`
- Added `## Options` and `## Function` section to the `README.md`
- Added new tests


## v0.1.1 - 2016-07-21

- Since the docs inside `__using__` were unreachable the design of application was changed
- Now `opts` keyword-list contains `:slug_build` parameter with a `build_slug` callback
- Also `coverage` information is updated, since more code is reachable now


## v0.1.0 - 2016-07-21

- Initial release


test/test_helper.exs
# Slug fields:

defmodule EctoAutoslugField.Test.TestSchema.SimpleSlug do
  use EctoAutoslugField.Slug, from: :name, to: :simple_slug
end

defmodule EctoAutoslugField.Test.TestSchema.SimpleSlugForce do
  use EctoAutoslugField.Slug, from: :name, to: :simple_slug_force
end

defmodule EctoAutoslugField.Test.TestSchema.MultipleSourcesSlug do
  use EctoAutoslugField.Slug,
    from: [:name, :company],
    to: :multiple_sources_slug
end

defmodule EctoAutoslugField.Test.TestSchema.IdFieldSlug do
  use EctoAutoslugField.Slug,
    from: [:name, :company, :link_id],
    to: :id_field_slug
end

defmodule EctoAutoslugField.Test.TestSchema.DateTimeSlug do
  use EctoAutoslugField.Slug,
    from: [
      :name,
      :company,
      :date_field,
      :time_field,
      :naive_datetime_field,
      :utc_datetime_field
    ],
    to: :datetime_slug
end

defmodule EctoAutoslugField.Test.TestSchema.MultiTypeSlug do
  use EctoAutoslugField.Slug,
    from: [:name, :company, :float_amount, :decimal_amount, :flag],
    to: :multitype_slug
end

defmodule EctoAutoslugField.Test.TestSchema.ComplexSlug do
  use EctoAutoslugField.Slug, to: :complex_slug

  def get_sources(_changeset, _opts), do: [:name, :company]

  def build_slug(sources, changeset) do
    sources
    |> super(changeset)
    |> String.replace("-", "+")
  end
end

defmodule EctoAutoslugField.Test.TestSchema.ConditionalSlug do
  use EctoAutoslugField.Slug, to: :conditional_slug

  def get_sources(changeset, _opts) do
    if Map.has_key?(changeset.changes(), :company) do
      [:name, :company]
    else
      [:name]
    end
  end
end

defmodule EctoAutoslugField.Test.TestSchema.AlwaysChangeSlug do
  use EctoAutoslugField.Slug,
    from: :name,
    to: :changing_slug,
    always_change: true
end

# Test schema:

defmodule EctoAutoslugField.Test.User do
  use Ecto.Schema

  import Ecto.Changeset

  # credo:disable-for-next-line Credo.Check.Readability.AliasOrder
  alias EctoAutoslugField.Test.TestSchema.{
    AlwaysChangeSlug,
    ConditionalSlug,
    ComplexSlug,
    DateTimeSlug,
    IdFieldSlug,
    MultiTypeSlug,
    MultipleSourcesSlug,
    SimpleSlugForce,
    SimpleSlug
  }

  schema "user" do
    field(:name, :string)
    field(:company, :string)
    field(:link_id, :integer)
    field(:float_amount, :float)
    field(:decimal_amount, :decimal)
    field(:flag, :boolean)
    field(:date_field, :date)
    field(:time_field, :time)
    field(:naive_datetime_field, :naive_datetime)
    field(:utc_datetime_field, :utc_datetime)

    field(:simple_slug, SimpleSlug.Type)
    field(:simple_slug_force, SimpleSlugForce.Type)
    field(:multiple_sources_slug, MultipleSourcesSlug.Type)
    field(:id_field_slug, IdFieldSlug.Type)
    field(:datetime_slug, DateTimeSlug.Type)
    field(:multitype_slug, MultiTypeSlug.Type)
    field(:complex_slug, ComplexSlug.Type)
    field(:conditional_slug, ConditionalSlug.Type)
    field(:changing_slug, AlwaysChangeSlug.Type)
  end

  def changeset(model, params \\ :invalid) do
    all_fields = [
      :name,
      :company,
      :link_id,
      :float_amount,
      :decimal_amount,
      :flag,
      :date_field,
      :time_field,
      :naive_datetime_field,
      :utc_datetime_field,
      :simple_slug,
      :simple_slug_force,
      :multiple_sources_slug,
      :id_field_slug,
      :datetime_slug,
      :multitype_slug,
      :complex_slug,
      :conditional_slug,
      :changing_slug
    ]

    model
    |> cast(params, all_fields)
    |> validate_required([:name])
    |> SimpleSlug.maybe_generate_slug()
    |> MultipleSourcesSlug.maybe_generate_slug()
    |> IdFieldSlug.maybe_generate_slug()
    |> DateTimeSlug.maybe_generate_slug()
    |> MultiTypeSlug.maybe_generate_slug()
    |> ComplexSlug.maybe_generate_slug()
    |> ConditionalSlug.maybe_generate_slug()
    |> AlwaysChangeSlug.maybe_generate_slug()
    |> SimpleSlugForce.force_generate_slug()
  end
end

# Run tests:
ExUnit.start()


test/ecto_autoslug_field_test/slug_test.exs
defmodule EctoAutoslugField.SlugTest do
  use ExUnit.Case
  alias EctoAutoslugField.Test.User

  @valid_attrs %{
    name: "Nikita Sobolev",
    company: "wemake.services",
    link_id: 1,
    date_field: ~N[2015-02-20 00:00:00],
    time_field: ~N[2000-01-01 19:50:07],
    naive_datetime_field: ~N[2015-01-21 23:34:07],
    utc_datetime_field: ~N[2015-11-03 03:14:07],
    float_amount: 2.31,
    decimal_amount: Decimal.from_float(4.765),
    flag: true
  }

  setup do
    {:ok,
     %{user: User.changeset(%User{simple_slug_force: "foo-bar"}, @valid_attrs)}}
  end

  test "changeset is valid", %{user: user} do
    assert user.valid?
  end

  test "simple slug", %{user: user} do
    assert user.changes.simple_slug == "nikita-sobolev"
  end

  test "simple slug force regeneration", %{user: user} do
    assert user.changes.simple_slug_force == "nikita-sobolev"
  end

  test "multiple sources slug", %{user: user} do
    assert user.changes.multiple_sources_slug ==
             "nikita-sobolev-wemake-services"
  end

  test "id field slug", %{user: user} do
    assert user.changes.id_field_slug == "nikita-sobolev-wemake-services-1"
  end

  test "datetime field slug", %{user: user} do
    assert user.changes.datetime_slug ==
             "nikita-sobolev-wemake-services-2015-02-20-19-50-07-2015-01-21-23-34-07-2015-11-03-03-14-07z"
  end

  test "multitype field slug", %{user: user} do
    assert user.changes.multitype_slug ==
             "nikita-sobolev-wemake-services-2-31-4-765-true"
  end

  test "multiple source when some are not set" do
    user = User.changeset(%User{}, %{name: "Nikita Sobolev"})
    assert user.changes.multiple_sources_slug == "nikita-sobolev"
  end

  test "complex slug", %{user: user} do
    assert user.changes.complex_slug == "nikita+sobolev+wemake+services"
  end

  test "conditional slug", %{user: user} do
    assert user.changes.conditional_slug == "nikita-sobolev-wemake-services"

    no_company = User.changeset(%User{}, %{name: "Nikita Sobolev"})
    assert no_company.changes.conditional_slug == "nikita-sobolev"
  end

  test "always changing slug" do
    user = %User{
      id: 1,
      changing_slug: "nikita-sobolev",
      name: "Nikita Sobolev"
    }

    changeset = User.changeset(user, %{name: "Sobolev Nikita"})
    assert changeset.changes.changing_slug == "sobolev-nikita"
  end

  test "basic slugs do not change" do
    user = %User{
      id: 1,
      name: "Nikita Sobolev",
      simple_slug: "nikita-sobolev",
      conditional_slug: "nikita-sobolev",
      complex_slug: "nikita+sobolev+wemake+services",
      multiple_sources_slug: "nikita-sobolev"
    }

    changeset = User.changeset(user, %{name: "Sobolev Nikita"})
    refute Map.has_key?(changeset.changes, :simple_slug)
    refute Map.has_key?(changeset.changes, :conditional_slug)
    refute Map.has_key?(changeset.changes, :multiple_sources_slug)
    refute Map.has_key?(changeset.changes, :complex_slug)
  end

  test "unique_constraint sets the constraint", %{user: user} do
    alias EctoAutoslugField.Test.TestSchema.SimpleSlug
    constrained = user |> SimpleSlug.unique_constraint()

    assert %{
             constraint: "user_simple_slug_index",
             field: :simple_slug,
             error_message: "has already been taken",
             error_type: :unique,
             type: :unique,
             match: :exact
           } in constrained.constraints
  end
end


test/ecto_autoslug_field_test/type_test.exs
defmodule EctoAutoslugField.TypeTest do
  use ExUnit.Case
  alias EctoAutoslugField.Type

  doctest EctoAutoslugField

  test "basic type" do
    assert Type.type() == :string
  end

  test "cast string" do
    assert Type.cast("value") == {:ok, "value"}
  end

  test "cast error" do
    assert Type.cast(:atom) == :error
  end

  test "load string" do
    assert Type.load("value") == {:ok, "value"}
  end

  test "load error" do
    assert Type.load([]) == :error
  end

  test "dump string" do
    assert Type.dump("value") == {:ok, "value"}
  end

  test "dump error" do
    assert Type.dump(%{}) == :error
  end
end


lib/ecto_autoslug_field/slug.ex
defmodule EctoAutoslugField.SlugBase do
  @moduledoc """
  This module defines all functions `Slug` module uses.
  """

  alias Ecto.Changeset
  alias EctoAutoslugField.SlugGenerator

  @doc """
  This function is used to generate slug.

  It is called 'maybe' since it may not generate slug for several reasons:

    1. It was already created and `:always_change` option was not set
    2. The source fields for the slug were empty

  This function takes changeset as an input and returns changeset.
  """
  @spec maybe_generate_slug(Changeset.t(), atom() | list() | none, Keyword.t()) ::
          Changeset.t()
  def maybe_generate_slug(changeset, sources, opts) do
    SlugGenerator.maybe_generate_slug(changeset, sources, opts)
  end

  @doc """
  This is just a helper function to check for uniqueness.

  It basically just wraps `Ecto.Changeset` and set a proper field name.
  """
  @spec unique_constraint(Changeset.t(), atom(), Keyword.t()) :: Changeset.t()
  def unique_constraint(changeset, to, opts \\ []) do
    Changeset.unique_constraint(changeset, to, opts)
  end

  @doc """
  This function is used to get sources for the slug.

  There can be many use cases when this behavior is required,
  here are the brief examples:

    1. Conditional slug sources
    2. Add any data from different sources

  This function should return `list` of atoms or binaries, or `nil`.

  When processing the returned list:

    1. `atom`-key is supposed to identify the model field
    2. `binary`-key is treated as a data itself, it won't be changed
  """
  @spec get_sources(Changeset.t(), Keyword.t()) :: list(atom() | binary())
  def get_sources(_changeset, [from: from] = _opts) do
    # This code is only used in macros, so it is not tracked by `coveralls`.
    # coveralls-ignore-start
    [from]
    # coveralls-ignore-stop
  end

  @doc """
  This function is used to build the slug itself.

  This function is a place to modify the result slug.
  For convenience you can call `super(sources, changeset)`
  which will return the slug binary.
  `super(sources)` uses [`slugify`](https://github.com/jayjun/slugify),
  but you can completely change slug-engine to your own.

  Note: this function will only be called if `sources` is not empty.
  Also important this function will be called only
  once for the normal workflow. And every time for `:always_change`.
  So you can do some heavy computations.

  If for some reason slug should not be set -
  just return `nil` or empty `binary`.

  It should return a `binary` or `nil`.
  """
  @spec build_slug(Keyword.t(), Changeset.t() | nil) :: String.t() | nil
  def build_slug(sources, changeset),
    do: SlugGenerator.build_slug(sources, changeset)
end

defmodule EctoAutoslugField.Slug do
  @moduledoc ~S"""
  This module defines all the required functions and modules to work with.

  ## Examples

  To create a simple 'Slug' field do:

      defmodule MyCustomSlug do
        use EctoAutoslugField.Slug, from: :name_field, to: :slug_field
      end

  It is also possible to override `get_sources/2` and `build_slug/2` functions
  which are part of the AutoslugField's API.

  More complex example with the optional sources
  and custom slug generation function:

      defmodule MyComplexSlug do
        use EctoAutoslugField.Slug, to: :slug_field

        def get_sources(changeset, _opts) do
          basic_fields = [:name, :surname]

          if is_company_info_set(changeset) do
            # We want to track changes in the person's company:
            basic_fields ++ [:company, :position]
          else
            basic_fields
          end
        end

        def build_slug(sources, changeset) do
          super(sources, changeset)  # Calls the `SlugGenerator.build_slug/1`
          |> String.replace("-", "+")
        end
      end

  It is also possible to always change your slug, even if it was already set:

      defmodule ThisSlugShouldChange do
        use EctoAutoslugField.Slug, from: :some_field,
          to: :slug_field, always_change: true
      end

  If you want to change slug for only one instance
  without setting `always_change` option, use `force_generate_slug/1` function:

      defmodule SimpleSlugForce do
        use EctoAutoslugField.Slug, from: :name, to: :some_slug
      end

  Then you can use `SimpleSlugForce.force_generate_slug(changeset)`
  for any instances, that needs to recreate slugs for some reason.

  Be careful with these options, since
  [cool URIs do not change](https://www.w3.org/Provider/Style/URI.html).
  """

  defmacro __using__(options) do
    caller = __CALLER__.module

    quote location: :keep, bind_quoted: [options: options, caller: caller] do
      alias EctoAutoslugField.SlugBase

      # Opts:

      @from Keyword.get(options, :from, nil)
      @to Keyword.get(options, :to, :slug)
      @always_change Keyword.get(options, :always_change, false)

      # Custom Type:

      defmodule Module.concat(caller, "Type") do
        @moduledoc """
        This module represent the auto-generated `Ecto.Type` for slug-field.

        It basically just calls the methods of the basic `Type` module.
        """
        @behaviour Ecto.Type

        alias EctoAutoslugField.Type

        def type, do: Type.type()
        def cast(value), do: Type.cast(value)
        def load(value), do: Type.load(value)
        def dump(value), do: Type.dump(value)
        def embed_as(value), do: Type.embed_as(value)
        def equal?(term1, term2), do: Type.equal?(term1, term2)
      end

      defp generate_slug_opts do
        [
          from: @from,
          to: @to,
          always_change: @always_change,
          slug_builder: &build_slug/2
        ]
      end

      defp generate_slug_sources(changeset, opts) do
        if opts[:from] == nil do
          get_sources(changeset, opts)
        else
          @from
        end
      end

      # Public functions:

      def maybe_generate_slug(changeset) do
        opts = generate_slug_opts()
        sources = generate_slug_sources(changeset, opts)

        SlugBase.maybe_generate_slug(changeset, sources, opts)
      end

      def force_generate_slug(changeset) do
        opts = generate_slug_opts() |> Keyword.put(:always_change, true)
        sources = generate_slug_sources(changeset, opts)

        SlugBase.maybe_generate_slug(changeset, sources, opts)
      end

      def unique_constraint(changeset, opts \\ []) do
        SlugBase.unique_constraint(changeset, @to, opts)
      end

      # Client API:

      def get_sources(changeset, opts) do
        SlugBase.get_sources(changeset, opts)
      end

      def build_slug(sources) do
        SlugBase.build_slug(sources, nil)
      end

      def build_slug(sources, _changeset) do
        build_slug(sources)
      end

      defoverridable get_sources: 2, build_slug: 2, build_slug: 1
    end
  end
end


lib/ecto_autoslug_field/slug_generator.ex
defmodule EctoAutoslugField.SlugGenerator do
  @moduledoc """
  This module works with slugs itself. It is just a wrapper around 'Slugify'.

  It is suited for inner use.
  """

  alias Ecto.Changeset

  import Ecto.Changeset,
    only: [
      put_change: 3,
      get_change: 3
    ]

  @doc """
  This is a public wrapper around `do_build_slug/1` functions.

  Default slug builder.
  """
  @spec build_slug(Keyword.t(), Changeset.t() | nil) :: String.t() | nil
  def build_slug(sources, _changeset), do: do_build_slug(sources)

  @doc """
  This function conditionally generates slug.

  This function prepares sources and then calls `do_generate_slug/3`.
  """
  @spec maybe_generate_slug(Changeset.t(), atom() | list(), Keyword.t()) ::
          Changeset.t()
  def maybe_generate_slug(changeset, source, opts) when is_atom(source) do
    source_value = get_field_data(changeset, source, opts)
    do_generate_slug(changeset, source_value, opts)
  end

  def maybe_generate_slug(changeset, sources, opts) do
    cleaned_sources =
      sources
      |> Enum.map(fn v -> get_field_data(changeset, v, opts) end)
      |> Enum.filter(fn v -> has_value?(v) end)

    do_generate_slug(changeset, cleaned_sources, opts)
  end

  # Private functions:

  defp do_build_slug(source) when is_binary(source) do
    source
    |> Slug.slugify()
  end

  defp do_build_slug(sources) do
    sources
    |> Enum.join("-")
    |> Slug.slugify()
  end

  defp do_generate_slug(changeset, sources, opts) do
    slug_key = Keyword.get(opts, :to)
    slug_builder = Keyword.get(opts, :slug_builder)
    always_change = Keyword.get(opts, :always_change, false)

    slug_field = Map.get(changeset.data, slug_key)

    if always_change == true or slug_field == nil do
      # We only generate slug on two occasions:
      # 1. `always_change` flag is set to true
      # 2. `slug_field` is not set, meaning it is a first-time-slug
      do_put_change(changeset, slug_key, slug_builder, sources)
    else
      changeset
    end
  end

  defp do_put_change(changeset, _, _, []), do: changeset
  defp do_put_change(changeset, _, _, nil), do: changeset

  defp do_put_change(changeset, slug_key, slug_builder, sources) do
    # `slug_builder` will be called only if the slug-building occasion
    # was met and the `sources` is not empty.
    slug_string = slug_builder.(sources, changeset)
    put_change(changeset, slug_key, slug_string)
  end

  defp get_field_data(changeset, source, opts) when is_atom(source) do
    always_change = Keyword.get(opts, :always_change, false)
    source_value = get_change(changeset, source, nil)

    if always_change do
      source_value || Map.get(changeset.data, source)
    else
      source_value
    end
  end

  defp get_field_data(_, source, _) when is_binary(source), do: source

  defp has_value?(nil), do: false
  defp has_value?(string) when is_binary(string), do: String.trim(string) != ""
  defp has_value?(_), do: true
end


mix.exs
defmodule EctoAutoslugField.Mixfile do
  use Mix.Project

  @source_url "https://github.com/sobolevn/ecto_autoslug_field"
  @version "3.1.0"

  def project do
    [
      app: :ecto_autoslug_field,
      version: @version,
      elixir: "~> 1.13",
      deps: deps(),
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,

      # Hex:
      docs: docs(),
      package: package(),

      # Test coverage:
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ],

      # Dialyzer:
      dialyzer: [
        plt_add_deps: :apps_direct,
        plt_add_apps: [:ecto]
      ]
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp deps do
    [
      {:ecto, ">= 3.7.0"},

      # Slugs:
      {:slugify, "~> 1.3"},

      # Testing:
      {:excoveralls, "~> 0.17", only: :test},
      {:castore, "~> 1.0", only: :test, runtime: false},
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 1.3", only: [:dev, :test], runtime: false},

      # Documentation:
      {:ex_doc, ">= 0.30.0", only: :dev, runtime: false}
    ]
  end

  defp docs do
    [
      extras: [
        "CHANGELOG.md",
        "LICENSE.md": [title: "License"],
        "README.md": [title: "Overview"]
      ],
      main: "readme",
      source_url: @source_url,
      homepage_url: @source_url,
      formatters: ["html"]
    ]
  end

  defp package do
    [
      description: "Autoslug field for Ecto.",
      maintainers: ["Nikita Sobolev"],
      licenses: ["MIT"],
      files: ~w(mix.exs README.md lib),
      links: %{
        "Changelog" => "https://hexdocs.pm/ecto_autoslug_field/changelog.html",
        "GitHub" => @source_url
      }
    ]
  end
end
